File: HackerRank/NewYearChaos/chaos.c
#include <assert.h>
#include <limits.h>
#include <math.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct bribe_mapping {
  int person;
  short count;
}

short* count_for_person(bribe_mapping *bribes, int size, int person) {
  for (size_t i = 0; i < size; i++) {
    if ((*bribes[i].person) == person) {
      return &bribes[i].count
    } else if ( (*bribes[i].person) == 0 ){
      return bribes[i] = &bribe_mapping{person, 0};
    }
  }
}

void minimum_bribes(int size, int* queue){
  int swaps = 0;
  bool swapped = false;
  bool keep_sorting = true;
  int temp_store;
  bribe_mapping* bribes = (bribe_mapping*) calloc(size, bribe_mapping);

  for (size_t q_pos = 0; q_pos < (size - 1); q_pos++) {
    if( keep_sorting ) {
      for (size_t sort_pos = 0; sort_pos < (size - 1); sort_pos++) {
        if( queue[sort_pos] > queue[sort_pos + 1] ) {
          short* count = count_for_person(bribes, size, queue[sort_pos]-1);
          (count*)++;
          if ( (*count) > 2 ) {
            puts("Too chaotic");
            return;
          }
        }
      }
    }
  }
}

File: HackerRank/NewYearChaos/chaos.py


CHAOS_ERR = "Too chaotic"
allowed_epsilon = 2

class TestCase:
    def __init__(self, bribes, *values):
        self.bribes = bribes or CHAOS_ERR
        self.values = list(values)

cases = (
    TestCase(3, 2, 1, 5, 3, 4),
    TestCase(None, 2, 5, 1, 3, 4),
    TestCase(None, 5, 1, 2, 3, 7, 8, 6, 4),
    TestCase(7, 1, 2, 5, 3, 7, 8, 6, 4)
)

def test_chaos():
    for case in cases:
        assert minimum_bribes(case.values) == (case.bribes)

def minimum_bribes(queue):
    swaps = 0
    swapped = False
    bribes = {}

    for q_pos in xrange(len(queue) - 1):
        for sort_pos in xrange(len(queue) - 1):
            if queue[sort_pos] > queue[sort_pos + 1]:
                bribes[queue[sort_pos]-1] = bribes.get(queue[sort_pos]-1, 0) + 1
                if bribes[queue[sort_pos]-1] > 2:
                    return CHAOS_ERR
                queue[sort_pos], queue[sort_pos + 1] = queue[sort_pos + 1], queue[sort_pos]
                swaps += 1
                swapped = True
        if not swapped:
            break
    return swaps
